src/modules/global_ban_protection/index.js

```diff
@@ -0,0 +1,61 @@
+const settings = require('../../settings');
+const twitch = require('../../utils/twitch');
+
+// Regular users are allowed 20 messages in 30 seconds, mods and higher
+// positions 100 messages in 30 seconds, see
+// https://discuss.dev.twitch.tv/t/message-limit-for-moderators/3042/4
+const regularMsgLimit = 20;
+const modMsgLimit = 100;
+const timeLimit = 30;
+
+class GlobalBanProtectionModule {
+    constructor() {
+        settings.add({
+            id: 'globalBanProtection',
+            name: 'Global Ban Protection',
+            description: 'Prevent sending a message if it would result in a global ban',
+            default: false,
+        });
+
+        this.msgLimit = -1;
+        this.msgTimeStamps = new Array();
+    }
+
+    onSendMessage(sendState) {
+        if (settings.get('globalBanProtection') === false) return;
+
+        if (this.msgLimit < 0) {
+            // Do this here for a reliable result, wasn't reliable in the constructor
+            if (twitch.getCurrentUserIsOwner() || twitch.getCurrentUserIsModerator()) {
+                this.msgLimit = modMsgLimit;
+            } else {
+                this.msgLimit = regularMsgLimit;
+            }
+        }
+
+        // Clear collected time stamps older than timeLimit seconds as they are
+        // no longer relevant for a global ban.
+        let done = false;
+        while (!done) {
+            if (this.msgTimeStamps.length === 0) break;
+
+            const oldestMsgTimeStamp = this.msgTimeStamps.shift();
+            if ((Date.now() - oldestMsgTimeStamp) / 1000 <= timeLimit) {
+                this.msgTimeStamps.unshift(oldestMsgTimeStamp);
+                done = true;
+            }
+        }
+
+        // At this point, this.msgTimeStamps.length is the number of messages
+        // sent in the last timeLimit seconds. If there is room for another
+        // message, collect its time stamp. Else, prevent the global ban.
+        if (this.msgTimeStamps.length < this.msgLimit) {
+            this.msgTimeStamps.push(Date.now());
```

Comment:
rate limits are generally not calculated like this, and I would argue this is not how Twitch is performing their calculation.

generally limits like this are windows of wall clock time. if you send your first message at 12:12:30 the rate limit expires at 12:13:00 if it's a 30s rate limit. you would want to store an expiration time when the first message is sent, and keep a count during that duration.

sample data model:

```
{
    expires: Date.now() + RATE_LIMIT_EXPIRATION,
    count: 0,
}
```